File: mobile/android/app/src/main/java/com/ttllegacy/api/ApiClient.kt
package com.ttllegacy.api

import com.ttllegacy.models.*
import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.engine.android.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.plugins.logging.*
import io.ktor.client.request.*
import io.ktor.http.*
import io.ktor.serialization.kotlinx.json.*
import kotlinx.serialization.json.Json
import javax.inject.Inject
import javax.inject.Singleton

sealed class ApiResult<out T> {
    data class Success<T>(val data: T) : ApiResult<T>()
    data class Error(val message: String, val code: Int = 0) : ApiResult<Nothing>()
    object NetworkUnavailable : ApiResult<Nothing>()
}

@Singleton
class ApiClient @Inject constructor(
    private val tokenProvider: TokenProvider,
    private val networkMonitor: NetworkMonitor,
    private val offlineCache: OfflineCache,
    private val baseUrl: String
) {
    private val client = HttpClient(Android) {
        install(ContentNegotiation) {
            json(Json { ignoreUnknownKeys = true; isLenient = true })
        }
        install(Logging) { level = LogLevel.INFO }
    }

    // Auth
    suspend fun getChallenge(): ApiResult<AuthChallenge> = get("/auth/challenge")
    suspend fun verifyPasskey(req: PasskeyVerifyRequest): ApiResult<AuthToken> = post("/auth/verify", req)
    suspend fun registerPasskey(req: PasskeyRegisterRequest): ApiResult<Unit> = post("/auth/register", req)

    // Vaults
    suspend fun listVaults(): ApiResult<List<Vault>> = get("/vaults")
    suspend fun getVault(id: String): ApiResult<Vault> = get("/vaults/$id")
    suspend fun createVault(req: CreateVaultRequest): ApiResult<Vault> = post("/vaults", req)
    suspend fun checkIn(vaultId: String): ApiResult<Unit> = post("/vaults/$vaultId/checkin", Unit)
    suspend fun deposit(vaultId: String, amount: Long): ApiResult<Vault> =
        post("/vaults/$vaultId/deposit", mapOf("amount" to amount))
    suspend fun withdraw(vaultId: String, amount: Long): ApiResult<Vault> =
        post("/vaults/$vaultId/withdraw", mapOf("amount" to amount))

    // Push
    suspend fun registerPushToken(token: String): ApiResult<Unit> =
        post("/notifications/register", PushRegistration(token = token))
    suspend fun unregisterPushToken(token: String): ApiResult<Unit> =
        delete("/notifications/register", PushRegistration(token = token))

    // Internals
    private suspend inline fun <reified T> get(path: String): ApiResult<T> {
        if (!networkMonitor.isConnected) {
            val cached = offlineCache.load(path)
            return if (cached != null) ApiResult.Success(Json.decodeFromString(cached))
            else ApiResult.NetworkUnavailable
        }
        return runCatching {
            val response = client.get("$baseUrl$path") { bearerAuth() }
            when (response.status.value) {
                in 200..299 -> {
                    val body: T = response.body()
                    offlineCache.save(path, Json.encodeToString(kotlinx.serialization.serializer(), body))
                    ApiResult.Success(body)
                }
                401 -> ApiResult.Error("Unauthorized", 401)
                404 -> ApiResult.Error("Not found", 404)
                else -> ApiResult.Error("Server error ${response.status.value}", response.status.value)
            }
        }.getOrElse { ApiResult.Error(it.message ?: "Unknown error") }
    }

    private suspend inline fun <reified B, reified T> post(path: String, body: B): ApiResult<T> {
        if (!networkMonitor.isConnected) return ApiResult.NetworkUnavailable
        return runCatching {
            val response = client.post("$baseUrl$path") {
                bearerAuth()
                contentType(ContentType.Application.Json)
                setBody(body)
            }
            when (response.status.value) {
                in 200..299 -> ApiResult.Success(if (T::class == Unit::class) Unit as T else response.body())
                401 -> ApiResult.Error("Unauthorized", 401)
                else -> ApiResult.Error("Server error ${response.status.value}", response.status.value)
            }
        }.getOrElse { ApiResult.Error(it.message ?: "Unknown error") }
    }

    private suspend inline fun <reified B, reified T> delete(path: String, body: B): ApiResult<T> {
        if (!networkMonitor.isConnected) return ApiResult.NetworkUnavailable
        return runCatching {
            val response = client.delete("$baseUrl$path") {
                bearerAuth()
                contentType(ContentType.Application.Json)
                setBody(body)
            }
            ApiResult.Success(if (T::class == Unit::class) Unit as T else response.body())
        }.getOrElse { ApiResult.Error(it.message ?: "Unknown error") }
    }

    private fun HttpRequestBuilder.bearerAuth() {
        tokenProvider.token?.let { header(HttpHeaders.Authorization, "Bearer $it") }
    }
}

File: mobile/ios/TTLLegacy/Sources/Services/APIClient.swift
import Foundation

enum APIError: LocalizedError {
    case unauthorized
    case notFound
    case serverError(String)
    case networkUnavailable
    case decodingFailed

    var errorDescription: String? {
        switch self {
        case .unauthorized:       return "Authentication required"
        case .notFound:           return "Resource not found"
        case .serverError(let m): return m
        case .networkUnavailable: return "No internet connection"
        case .decodingFailed:     return "Invalid server response"
        }
    }
}

final class APIClient {
    static let shared = APIClient()

    private let baseURL: URL
    private let session: URLSession
    private let decoder: JSONDecoder

    private init() {
        let urlString = Bundle.main.object(forInfoDictionaryKey: "API_BASE_URL") as? String
            ?? "https://api.ttl-legacy.app/v1"
        baseURL = URL(string: urlString)!
        session = URLSession(configuration: .default)
        decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        decoder.dateDecodingStrategy = .iso8601
    }

    // MARK: - Auth

    func getChallenge() async throws -> AuthChallenge {
        try await post(path: "/auth/challenge", body: EmptyBody())
    }

    func verifyPasskey(credentialID: String, clientDataJSON: String, signature: String) async throws -> AuthToken {
        let body = ["credential_id": credentialID,
                    "client_data_json": clientDataJSON,
                    "signature": signature]
        return try await post(path: "/auth/verify", body: body)
    }

    func registerPasskey(credentialID: String, publicKey: String, clientDataJSON: String) async throws {
        let body = ["credential_id": credentialID,
                    "public_key": publicKey,
                    "client_data_json": clientDataJSON]
        let _: EmptyBody = try await post(path: "/auth/register", body: body)
    }

    // MARK: - Vaults

    func listVaults() async throws -> [Vault] {
        try await get(path: "/vaults")
    }

    func getVault(id: String) async throws -> Vault {
        try await get(path: "/vaults/\(id)")
    }

    func createVault(beneficiary: String, checkInInterval: UInt64) async throws -> Vault {
        let body = ["beneficiary": beneficiary, "check_in_interval": checkInInterval] as [String: Any]
        return try await post(path: "/vaults", body: body)
    }

    func checkIn(vaultID: String) async throws {
        let _: EmptyBody = try await post(path: "/vaults/\(vaultID)/checkin", body: EmptyBody())
    }

    func deposit(vaultID: String, amount: Int64) async throws -> Vault {
        try await post(path: "/vaults/\(vaultID)/deposit", body: ["amount": amount])
    }

    func withdraw(vaultID: String, amount: Int64) async throws -> Vault {
        try await post(path: "/vaults/\(vaultID)/withdraw", body: ["amount": amount])
    }

    func getTTL(vaultID: String) async throws -> UInt64 {
        let result: [String: UInt64] = try await get(path: "/vaults/\(vaultID)/ttl")
        return result["ttl_remaining"] ?? 0
    }

    // MARK: - Push Notifications

    func registerPushToken(_ token: String) async throws {
        let body = PushRegistration(token: token, platform: "ios")
        let _: EmptyBody = try await post(path: "/notifications/register", body: body)
    }

    func unregisterPushToken(_ token: String) async throws {
        var req = request(path: "/notifications/register")
        req.httpMethod = "DELETE"
        req.httpBody = try? JSONEncoder().encode(PushRegistration(token: token, platform: "ios"))
        _ = try await execute(req)
    }

    // MARK: - Private helpers

    private func get<T: Decodable>(path: String) async throws -> T {
        var req = request(path: path)
        req.httpMethod = "GET"
        let data = try await execute(req)
        return try decode(data)
    }

    private func post<B: Encodable, T: Decodable>(path: String, body: B) async throws -> T {
        var req = request(path: path)
        req.httpMethod = "POST"
        req.httpBody = try JSONEncoder().encode(body)
        let data = try await execute(req)
        return try decode(data)
    }

    private func request(path: String) -> URLRequest {
        var req = URLRequest(url: baseURL.appendingPathComponent(path))
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let token = KeychainService.shared.loadToken() {
            req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        return req
    }

    private func execute(_ request: URLRequest) async throws -> Data {
        guard NetworkMonitor.shared.isConnected else {
            // Return cached data if available
            if let cached = OfflineCache.shared.load(for: request.url?.absoluteString ?? "") {
                return cached
            }
            throw APIError.networkUnavailable
        }
        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else { throw APIError.serverError("Invalid response") }
        switch http.statusCode {
        case 200...299:
            OfflineCache.shared.save(data, for: request.url?.absoluteString ?? "")
            return data
        case 401: throw APIError.unauthorized
        case 404: throw APIError.notFound
        default:
            let msg = (try? JSONDecoder().decode([String: String].self, from: data))?["error"] ?? "Server error"
            throw APIError.serverError(msg)
        }
    }

    private func decode<T: Decodable>(_ data: Data) throws -> T {
        if T.self == EmptyBody.self { return EmptyBody() as! T }
        do { return try decoder.decode(T.self, from: data) }
        catch { throw APIError.decodingFailed }
    }
}

private struct EmptyBody: Codable {}
